### DISCUSSION TITLE
why some code use ptx?

### DISCUSSION THREAD

**Discussion by MonroeD** [NVIDIA/cccl#1066] [CUB]
for some point, the performance of inline ptx code maybe not better than cuda c code

**Comment by jrhemstad** [comment]
Could you elaborate on your question? I believe you are asking why does CUB sometimes use inline PTX?

The answer is because sometimes that is the tool we need to get the job done. PTX is a very precise tool that allows us to say exactly what we want to happen and sometimes that's what we need.

**Reply by MonroeD** [reply]
e.g. this file https://github.com/NVIDIA/cub/blob/e882acc1e709feeb3c96f6f1d8a9fb7af9667823/cub/thread/thread_store.cuh
#define _CUB_STORE_16(cub_modifier, ptx_modifier)                                            \
    template<>                                                                              \
    __device__ __forceinline__ void ThreadStore<cub_modifier, uint4*, uint4>(uint4* ptr, uint4 val)                         \
    {                                                                                       \
        asm volatile ("st."#ptx_modifier".v4.u32 [%0], {%1, %2, %3, %4};" : :               \
            _CUB_ASM_PTR_(ptr),                                                             \
            "r"(val.x),                                                                     \
            "r"(val.y),                                                                     \
            "r"(val.z),                                                                     \
            "r"(val.w));                                                                    \
    }                                                                                       \
    template<>                                                                              \
    __device__ __forceinline__ void ThreadStore<cub_modifier, ulonglong2*, ulonglong2>(ulonglong2* ptr, ulonglong2 val)     \
    {                                                                                       \
        asm volatile ("st."#ptx_modifier".v2.u64 [%0], {%1, %2};" : :                       \
            _CUB_ASM_PTR_(ptr),                                                             \
            "l"(val.x),                                                                     \
            "l"(val.y));                                                                    \
    }

can we write follow bellow:

#define _CUB_STORE_16(cub_modifier, ptx_modifier)                                            \
    template<>                                                                              \
    __device__ __forceinline__ void ThreadStore<cub_modifier, uint4*, uint4>(uint4* ptr, uint4 val)                         \
    {                                                                                       \
             ptr[0].x = val.x;  \
             ptr[0].y = val.y;  \
             ptr[0].z = val.z;  \
             ptr[0].x = val.w; \
    }                                                                                       \
    template<>                                                                              \
    __device__ __forceinline__ void ThreadStore<cub_modifier, ulonglong2*, ulonglong2>(ulonglong2* ptr, ulonglong2 val)     \
    {                                                                                       \
           ptr[0].x = val.x; \ 
           ptr[0].y = val.y; \
    }

**Reply by jrhemstad** [reply]
You _could_, but you would lose control over the caching behavior of the store operation, see https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#cache-operators 

Things like this are what I meant about PTX being a precise tool that gives you more control than CUDA C++ code. These are very low level details that typical user code doesn't need to worry about. Our goal in CUB is to abstract away these details so you don't have to worry about them.
